App crashes on clicking Sensor Quick View tile when device isn't connected

**Actual Behaviour**

App crashes when device isn't connected.

**Expected Behaviour**

App shouldn't crash when device isn't connected instead user should be provided with appropriate message.

**Steps to reproduce it**

NavDrawer -> Application Fragment-> Sensor Quick View

**LogCat for the issue**

```
                                                                    Process: org.fossasia.pslab, PID: 21939
                                                                    java.lang.RuntimeException: Unable to start activity ComponentInfo{org.fossasia.pslab/org.fossasia.pslab.activity.SensorActivity}: java.lang.NullPointerException: Attempt to invoke virtual method 'java.util.ArrayList org.fossasia.pslab.communication.peripherals.I2C.scan(java.lang.Integer)' on a null object reference
                                                                        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2449)
                                                                        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2509)
                                                                        at android.app.ActivityThread.access$1000(ActivityThread.java:153)
                                                                        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1373)
                                                                        at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                        at android.os.Looper.loop(Looper.java:154)
                                                                        at android.app.ActivityThread.main(ActivityThread.java:5529)
                                                                        at java.lang.reflect.Method.invoke(Native Method)
                                                                        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:739)
                                                                        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:629)
                                                                     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.util.ArrayList org.fossasia.pslab.communication.peripherals.I2C.scan(java.lang.Integer)' on a null object reference
                                                                        at org.fossasia.pslab.sensorfragment.SensorFragmentMain.onCreate(SensorFragmentMain.java:66)
                                                                        at android.support.v4.app.Fragment.performCreate(Fragment.java:2180)
                                                                        at android.support.v4.app.FragmentManagerImpl.moveToState(FragmentManager.java:1244)
                                                                        at android.support.v4.app.FragmentTransition.addToFirstInLastOut(FragmentTransition.java:1085)
                                                                        at android.support.v4.app.FragmentTransition.calculateFragments(FragmentTransition.java:976)
                                                                        at android.support.v4.app.FragmentTransition.startTransitions(FragmentTransition.java:95)
                                                                        at android.support.v4.app.FragmentManagerImpl.executeOpsTogether(FragmentManager.java:2146)
                                                                        at android.support.v4.app.FragmentManagerImpl.optimizeAndExecuteOps(FragmentManager.java:2103)
                                                                        at android.support.v4.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:2013)
                                                                        at android.support.v4.app.FragmentController.execPendingActions(FragmentController.java:388)
                                                                        at android.support.v4.app.FragmentActivity.onStart(FragmentActivity.java:607)
                                                                        at android.support.v7.app.AppCompatActivity.onStart(AppCompatActivity.java:178)
                                                                        at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1238)
                                                                        at android.app.Activity.performStart(Activity.java:6320)
                                                                        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2412)
                                                                        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2509) 
                                                                        at android.app.ActivityThread.access$1000(ActivityThread.java:153) 
                                                                        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1373) 
                                                                        at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                        at android.os.Looper.loop(Looper.java:154) 
                                                                        at android.app.ActivityThread.main(ActivityThread.java:5529) 
                                                                        at java.lang.reflect.Method.invoke(Native Method)
```

**Screenshots of the issue**

NA

**Would you like to work on the issue?**

Yes

--- a/app/src/main/java/org/fossasia/pslab/activity/SensorActivity.java
+++ b/app/src/main/java/org/fossasia/pslab/activity/SensorActivity.java
@@ -42,9 +42,6 @@ public class SensorActivity extends AppCompatActivity {
         timegapLabel = (TextView) findViewById(R.id.tv_timegap_label);
         playPauseButton = (ImageButton) findViewById(R.id.imageButton_play_pause_sensor);
         play = false;
-
-
-        sensorDock.setVisibility(View.INVISIBLE);
         playPauseButton.setOnClickListener(new View.OnClickListener() {
             @Override
             public void onClick(View v) {
@@ -57,6 +54,7 @@ public class SensorActivity extends AppCompatActivity {
                 }
             }
         });
+        sensorDock.setVisibility(View.GONE);
 
         FragmentManager fragmentManager = getSupportFragmentManager();
         FragmentTransaction transaction = fragmentManager.beginTransaction();

--- a/app/src/main/java/org/fossasia/pslab/sensorfragment/SensorFragmentMain.java
+++ b/app/src/main/java/org/fossasia/pslab/sensorfragment/SensorFragmentMain.java
@@ -1,7 +1,10 @@
 package org.fossasia.pslab.sensorfragment;
 
+import android.graphics.Color;
 import android.os.Bundle;
 import android.support.annotation.Nullable;
+import android.support.design.widget.CoordinatorLayout;
+import android.support.design.widget.Snackbar;
 import android.support.v4.app.Fragment;
 import android.support.v4.app.FragmentManager;
 import android.support.v4.app.FragmentTransaction;
@@ -22,6 +25,7 @@ import org.fossasia.pslab.others.ScienceLabCommon;
 import java.io.IOException;
 import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.LinkedHashMap;
 
 /**
  * Created by asitava on 13/7/17.
@@ -31,7 +35,7 @@ public class SensorFragmentMain extends Fragment {
 
     private I2C i2c;
     private ScienceLab scienceLab;
-    private HashMap<Integer, String> sensorAddr = new HashMap<>();
+    private LinkedHashMap<Integer, String> sensorAddr = new LinkedHashMap<>();
     private ArrayList<Integer> data = new ArrayList<>();
     private ArrayList<String> dataAddress = new ArrayList<>();
     private ArrayList<String> dataName = new ArrayList<>();
@@ -50,49 +54,51 @@ public class SensorFragmentMain extends Fragment {
     public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         scienceLab = ScienceLabCommon.scienceLab;
-        i2c = scienceLab.i2c;
 
-        sensorAddr.put(0x60, "MCP4728");
-        sensorAddr.put(0x48, "ADS1115");
-        sensorAddr.put(0x23, "BH1750");
-        sensorAddr.put(0x77, "BMP180");
-        sensorAddr.put(0x5A, "MLX90614");
-        sensorAddr.put(0x1E, "HMC5883L");
-        sensorAddr.put(0x68, "MPU6050");
-        sensorAddr.put(0x40, "SHT21");
-        sensorAddr.put(0x39, "TSL2561");
+        if (scienceLab.isConnected()) {
+            i2c = scienceLab.i2c;
+            sensorAddr.put(0x60, "MCP4728");
+            sensorAddr.put(0x48, "ADS1115");
+            sensorAddr.put(0x23, "BH1750");
+            sensorAddr.put(0x77, "BMP180");
+            sensorAddr.put(0x5A, "MLX90614");
+            sensorAddr.put(0x1E, "HMC5883L");
+            sensorAddr.put(0x68, "MPU6050");
+            sensorAddr.put(0x40, "SHT21");
+            sensorAddr.put(0x39, "TSL2561");
 
-        try {
-            data = i2c.scan(null);
-        } catch (IOException e) {
-            e.printStackTrace();
-        }
-        if (data != null) {
-            for (Integer myInt : data) {
-                if (sensorAddr.get(myInt) != null) {
-                    dataAddress.add(String.valueOf(myInt));
-                    dataName.add(sensorAddr.get(myInt));
+            try {
+                data = i2c.scan(null);
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+            if (data != null) {
+                for (Integer myInt : data) {
+                    if (sensorAddr.get(myInt) != null) {
+                        dataAddress.add(String.valueOf(myInt));
+                        dataName.add(sensorAddr.get(myInt));
+                    }
+                }
+                for (String s : dataAddress) {
+                    tvData += s + ":" + sensorAddr.get(Integer.parseInt(s)) + "\n";
                 }
             }
-            for (String s : dataAddress) {
-                tvData += s + ":" + sensorAddr.get(Integer.parseInt(s)) + "\n";
-            }
+            String[] dataDisp = dataName.toArray(new String[dataName.size()]);
+            adapter = new ArrayAdapter<>(getContext(), R.layout.sensor_list_item, R.id.tv_sensor_list_item, dataDisp);
         }
-        String[] dataDisp = dataName.toArray(new String[dataName.size()]);
-        adapter = new ArrayAdapter<>(getContext(), R.layout.sensor_list_item, R.id.tv_sensor_list_item, dataDisp);
     }
 
     @Override
     public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
         View view = inflater.inflate(R.layout.sensor_main, container, false);
-
-        Button buttonSensorAutoscan = (Button) view.findViewById(R.id.button_sensor_autoscan);
+        final CoordinatorLayout coordinatorLayout = (CoordinatorLayout) view.findViewById(R.id.layout_container);
+        Button buttonSensorAutoScan = (Button) view.findViewById(R.id.button_sensor_autoscan);
         tvSensorScan = (TextView) view.findViewById(R.id.tv_sensor_scan);
         tvSensorScan.setText(tvData);
         lvSensor = (ListView) view.findViewById(R.id.lv_sensor);
         lvSensor.setAdapter(adapter);
 
-        buttonSensorAutoscan.setOnClickListener(new View.OnClickListener() {
+        buttonSensorAutoScan.setOnClickListener(new View.OnClickListener() {
             @Override
             public void onClick(View v) {
                 if (scienceLab.isConnected()) {
@@ -104,7 +110,7 @@ public class SensorFragmentMain extends Fragment {
 
                     if (data != null) {
                         for (Integer myInt : data) {
-                            if (myInt != null && sensorAddr.get(myInt) != null ) {
+                            if (myInt != null && sensorAddr.get(myInt) != null) {
                                 dataAddress.add(String.valueOf(myInt));
                                 dataName.add(sensorAddr.get(myInt));
                             }
@@ -119,6 +125,12 @@ public class SensorFragmentMain extends Fragment {
                     String[] dataDisp = dataName.toArray(new String[dataName.size()]);
                     adapter = new ArrayAdapter<>(getContext(), R.layout.sensor_list_item, R.id.tv_sensor_list_item, dataDisp);
                     lvSensor.setAdapter(adapter);
+                } else {
+                    Snackbar snackbar = Snackbar.make(coordinatorLayout, "Device not connected", Snackbar.LENGTH_SHORT);
+                    View snackBarView = snackbar.getView();
+                    TextView snackbarTextView = (TextView) snackBarView.findViewById(android.support.design.R.id.snackbar_text);
+                    snackbarTextView.setTextColor(Color.YELLOW);
+                    snackbar.show();
                 }
             }
         });

--- a/app/src/main/res/layout/activity_sensor.xml
+++ b/app/src/main/res/layout/activity_sensor.xml
@@ -2,7 +2,7 @@
 <RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:app="http://schemas.android.com/apk/res-auto"
     android:layout_width="match_parent"
-    android:layout_height="wrap_content"
+    android:layout_height="match_parent"
     android:focusable="true"
     android:focusableInTouchMode="true"
     app:layout_behavior="@string/appbar_scrolling_view_behavior">
@@ -41,6 +41,7 @@
             android:layout_height="wrap_content"
             android:layout_alignParentTop="true"
             android:layout_marginLeft="5dp"
+            android:layout_marginStart="5dp"
             android:layout_toEndOf="@+id/imageButton_play_pause_sensor"
             android:layout_toLeftOf="@+id/tv_timegap_label"
             android:layout_toRightOf="@+id/imageButton_play_pause_sensor"

--- a/app/src/main/res/layout/sensor_main.xml
+++ b/app/src/main/res/layout/sensor_main.xml
@@ -1,11 +1,14 @@
 <?xml version="1.0" encoding="utf-8"?>
-<ScrollView xmlns:android="http://schemas.android.com/apk/res/android"
+<android.support.design.widget.CoordinatorLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:id="@+id/layout_container"
     android:layout_width="match_parent"
-    android:layout_height="match_parent">
+    android:layout_height="match_parent"
+    android:fitsSystemWindows="true"
+    android:orientation="vertical">
 
     <LinearLayout
         android:layout_width="match_parent"
-        android:layout_height="wrap_content"
+        android:layout_height="match_parent"
         android:layout_marginEnd="10dp"
         android:layout_marginLeft="10dp"
         android:layout_marginRight="10dp"
@@ -39,10 +42,10 @@
                 android:layout_width="match_parent"
                 android:layout_height="match_parent"
                 android:background="@drawable/control_custom_border"
-                android:textSize="@dimen/sensor_body_font"
+                android:gravity="center"
                 android:textAlignment="center"
-                android:textStyle="bold"
-                android:gravity="center"/>
+                android:textSize="@dimen/sensor_body_font"
+                android:textStyle="bold" />
         </android.support.v4.widget.NestedScrollView>
 
         <TextView
@@ -59,8 +62,7 @@
         <ListView
             android:id="@+id/lv_sensor"
             android:layout_width="match_parent"
-            android:layout_height="200dp"
-            android:layout_marginBottom="10dp"
+            android:layout_height="match_parent"
             android:layout_marginTop="10dp" />
     </LinearLayout>
-</ScrollView>
+</android.support.design.widget.CoordinatorLayout>

